cobalt/dom/same_thread_media_source_attachment.cc

```diff
@@ -0,0 +1,144 @@
+      task_runner_(base::SequencedTaskRunner::GetCurrentDefault()) {}
+
+void SameThreadMediaSourceAttachment::TraceMembers(script::Tracer* tracer) {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+
+  tracer->Trace(attached_element_);
+  tracer->Trace(media_source_);
+}
+
+bool SameThreadMediaSourceAttachment::StartAttachingToMediaElement(
+    HTMLMediaElement* media_element) {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+
+  if (attached_element_) {
+    return false;
+  }
+  attached_element_ = base::AsWeakPtr(media_element);
+
+  bool result = media_source_->StartAttachingToMediaElement(this);
+
+  // Let the result of MediaSource::StartAttachingToMediaElement dictate if
+  // the reference to the HTMLMediaElement should be retained.
+  if (!result) {
+    attached_element_.reset();
+  }
+
+  return result;
+}
+
+void SameThreadMediaSourceAttachment::CompleteAttachingToMediaElement(
+    ::media::ChunkDemuxer* chunk_demuxer) {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+
+  media_source_->CompleteAttachingToMediaElement(chunk_demuxer);
+}
+
+void SameThreadMediaSourceAttachment::Close() {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+
+  media_source_->Close();
+}
+
+scoped_refptr<TimeRanges> SameThreadMediaSourceAttachment::GetBufferedRange()
+    const {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+
+  return media_source_->GetBufferedRange();
+}
+
+MediaSourceReadyState SameThreadMediaSourceAttachment::GetReadyState() const {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+
+  return media_source_->ready_state();
+}
+
+void SameThreadMediaSourceAttachment::DurationChanged(double duration) {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+  DCHECK(attached_element_);
+
+  bool request_seek = attached_element_->current_time(NULL) > duration;
+  attached_element_->DurationChanged(duration, request_seek);
+}
+
+bool SameThreadMediaSourceAttachment::HasMaxVideoCapabilities() const {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+  DCHECK(attached_element_);
+
+  return attached_element_->HasMaxVideoCapabilities();
+}
+
+double SameThreadMediaSourceAttachment::GetRecentMediaTime() {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+  DCHECK(attached_element_);
+
+  return attached_element_->current_time(NULL);
+}
+
+scoped_refptr<TimeRanges>
+SameThreadMediaSourceAttachment::GetElementBufferedRange() const {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+  DCHECK(attached_element_);
+
+  return attached_element_->buffered();
+}
+
+bool SameThreadMediaSourceAttachment::GetElementError() {
+  DCHECK_EQ(task_runner_, base::SequencedTaskRunner::GetCurrentDefault());
+  DCHECK(attached_element_);
+
+  return !!attached_element_->error();
```

Comment:
Do these `!!` have a special purpose?

Reply:
They were used to explicitly cast from a `scoped_refptr<MediaError>` to a `bool`.

I've changed this to use `static_cast` instead to be clearer in intention.